drivers/pwm/pwm_sam.c

```diff
@@ -37,6 +37,13 @@ static int sam_pwm_get_cycles_per_sec(const struct device *dev, uint32_t pwm,
 	return 0;
 }
 
+/* The SAMV71 HALs change the name of the field, so we need to
+ * define it this way to match how the other SoC variants name it
+ */
+#if defined(CONFIG_SOC_SERIES_SAMV71)
+#define PWM_CH_NUM PwmChNum
+#endif
```

Comment:
In a PR which I've just submitted #zephyrproject-rtos/hal_atmel/pull/15 (I've noticed this one too late) I've patched the HAL rather than the driver. I'm not sure which approach is better but we probably should align on one. It's unfortunate that Atmel HAL headers are low quality.

Reply:
Noting that there will be too many patches required to support all SAM family devices (i.e. `sam3*`, `sam4*`, `sam7*`), I wonder if we should just do [what we did for the SAM0 family](https://github.com/zephyrproject-rtos/zephyr/blob/ce003ffa458586a830b8d861b2b2ae580b6658ef/soc/arm/atmel_sam0/common/sercom_fixup_samd5x.h).

Reply:
@mnkp I was thinking that since we don't have a clear way of getting fixes back upstream like we would for STM or NXP devices doing something like what @stephanosio is suggesting would make sense.

Reply:
I take look at all SAM3/4/7 PWM and it is almost same IP block. There are differences related to DMA and others small improvements on some SoC. I believe if we follow SAM4S datasheet (minimal implementation), everything that we made will works for all others. Any other improvement can be easy added after. DMA is handled by different controllers: PDC or XDMAC. In this case, DMA need to be handled with different registers. The SAM4L must be excluded at Kconfig because don't have PWM feature.

My preference is direct adjust Atmel headers to make code easy to read and avoid #if blocks. Otherwise we start to create "garbage" on Zephyr tree and people can think that right thing to do. I don't have objections to @stephanosio suggestion but I think that king adjust should be made at hal_atmel, if majority think that is the right path.

Conflicts:

Only on SAM4E/SAME7x/V7x
0x24 PWM DMA Register
0xA0 PWM Spread Spectrum Register
0xA4 PWM Spread Spectrum Update Register
0xC0 PWM Fault Protection Value 2 Register

Only on SAM4E/SAM3X
0x100–0x128 Reserved for PDC registers

For any other, above is reserved.

Reply:
I also would prefer to avoid #if blocks in the driver. They do make code less readable. For me patching hal_atmel or following suggestion from @stephanosio would be a better choice.

Some years ago I opened a ticket with Atmel requesting fixes to some of the header files. There were quite a few issues. About the same time I was contacted by an Atmel engineer here on GitLab who was working on improving their header generation flow. He was friendly, helpful and aware that they need to do a better job. But at the end only half of the problems were fixed. And the headers are still low quality. It doesn't look like things changed much under control of Microchip.

Reply:
I don't expect MCHP will fix headers. In general, Manufacturers prefer invest on new things once current it is working. I already found a bug in the TWIM driver for SAM4L years ago. After a year or more, they fixed. Since we only need harmonize, I think it is better fix by demand, like in this case. The fix don't need be by a batch patch for all SoC at same time.

My preference is fixing on hal_atmel headers because we have a better history to tell e avoid complicated and dedicated zephyr headers, over time.

Thanks @galak for add #36187.

Reply:
Going to discuss the general topic of how we should handle HAL fixes when there is no clear upstream process for that HAL in the Process Meeting.

Reply:
Discussed briefly in the Process Meeting: the recommendation is to use the approach proposed above in this thread by @stephanosio , using the fixup files in the Zephyr main tree. We should capture this in the project documentation.

Reply:
@stephanosio @nandojve any suggestions on how I should move the "PWM" fixup over into the soc?

Reply:
Hi @galak ,

I believe you should add a fixup like pointed by @stephanosio to drop the below fragment:

```
#if defined(CONFIG_SOC_SERIES_SAMV71)
#define PWM_CH_NUM PwmChNum
#endif
```

at
`zephyr/soc/arm/atmel_sam/common/<driver>_fixup.h`

We work with this approach on a incremental way, so ,you can only fix the current interest. No need to fixup everything.
This allows us to harmonize SAM/SAM0 drivers fast. Later anyone can rework and move to hal_atmel.

Reply:
Updated with pwm_fixup.h